### ISSUE TITLE
Allow services to be provided asynchronously

### ISSUE THREAD

**Issue by dchambers** [BladeRunnerJS/brjs#1128]
At present, BRJS doesn't support services that need to be provided asynchronously. The [New Registry Proposal page](https://github.com/BladeRunnerJS/brjs/wiki/New-Registry-Proposal) contains a proposal that aims to address this, and which has taken consideration all of the feedback in this page.

**Comment by dchambers** [comment]
I've updated this issue to account for the impending ES6 problem.

**Comment by dchambers** [comment]
@adamshone / @andyberry88  / @briandipalma / @janhancic / @james-shaw-turner / @stevesouth : I've updated this issue based on feedback from @bit-shifter & @hobojed, and am now seeking wider feedback before we consider changing how we do services in BRJS.

@briandipalma, I can no longer remember what your remaining objections where to the proposal when we last discussed it, but if you still think this is completely the wrong approach can you please create a counter proposal. Or, if you just think this requires minor tweaks, then it would make more sense to comment directly in-line.

Thanks everyone!

**Comment by janhancic** [comment]
Just to clarify, the first comment in this issue is your current up-to-date proposal?

**Comment by dchambers** [comment]
Yep, that's correct. All pertinent information is in the first comment.

**Comment by adamshone** [comment]
Might be worth splitting this into two issues:

* The problem of asynchronously resolved services / delayed readiness services / services that depend on other services (unless I'm missing some subtleties I think all of those are just different descriptions for the same problem).
* The problem of how to swap out modules that have been required when running tests, given that the module system can cache the result of `require` calls.

**Comment by dchambers** [comment]
Yes, that may be a good thing to do from an administrative point of view, but since the design is affected by the complete set of use cases, I prefer to deal with all of it in one place from a design point of view. As it happens, this issue is also solving #1335 too.

**Comment by adamshone** [comment]
OK. I'm interested in the second problem because it affects testing for everything, not just services. Do you have a strawman code example for a unit test that uses this new mechanism? E.g if the class under test is:

````javascript
var logger = require('br/logger');

function Foo() {
   logger.info('constructed');
}

module.exports = Foo;
````

And the test is something like:

````javascript
var logger = /* a mock of */ require('br/logger');
var Foo = require('/foo');   /* but I want to inject the mock logger */

var myFoo = new Foo();
verify(logger.info.withArgs('constructed');
````

This is the basic test situation that the solution must address, we can also add services to make it more complicated if the solution covers this case.

**Comment by dchambers** [comment]
@adamshone, okay, then I think I misunderstood your point 2 then, as this issue has never been about allowing require paths to be mocked out. @janhancic created a similar and even more broad reaching issue for this (#1149), but this just seems to be more of the same kind of thing that @briandipalma is warning us not do due to the forthcoming ES6 module compatibility problems we'll face.

At present, the [sub-realms](https://github.com/BladeRunnerJS/browser-modules#sub-realms) feature of [browser-modules](https://github.com/BladeRunnerJS/browser-modules) is the only way to do this but, you guessed it, this will also be incompatible with ES6 modules! I'm guessing we either ignore ES6 modules, or start advocating explicit dependency injection in preference to using `require()` as a convenient _seam_ for performing IoC.

**Comment by briandipalma** [comment]
Hello,

Good discussion. I have a few questions. I'm not too clear on the introduction, the problems that we currently have part. Why is 1 a problem? The interface doesn't have to reflect the fact the implementation is provided async, that seems to be an implementation detail. How does it cause problems? 2 sounds like a bug, can we not just fix it? Neither issue is clear to me so I'm probably missing something.

I'd recommend against exporting instances as the default export, in effect they become singletons and you probably want to test the service class, maybe a well known symbol on the default export i.e. `INSTANCE` might be more appropriate?

```javascript
module.exports = ConsoleLogService;
module.exports.INSTANCE = Promise.resolve(new ConsoleLogService());
```

A similar pattern could be applied to remove the need for the double modules for `ServiceRegistry` and `ServiceRegistryClass` and other comparable instances.

We could consider using a type base registry in preference to a string based one at this juncture. It would allow tooling support to provide typing support once we move to using something like Flow or TS.

```javascript
var ServiceRegistry = require('br/ServiceRegistry');
var ConfigService = require('caplin/services/ConfigService');

var configService = ServiceRegistry.getService(ConfigService);
```

This would allow us to exploit polymorphism in the typed JS dialects. Strict nominal typing might be problematic in popouts (cross realm checks failing) so we might want to have a duck typing fallback if a service type doesn't match. Although I realize the benefit of this might be quite minor and not worth the hassle so we can just continue with the string based approach.

I guess that detecting when and where it is safe to convert libraries to use async registry access will be very tricky and would counsel against such changes, especially as I'm doubtful of the value of such conversions. Also debugging Promises when they fail can be annoying, although the dev tools are improving. Given that we might want to question supporting that extension.

I like the prefetch approach, it's unclear how the list of async provided services is kept in sync, I guess application developers will have to create and maintain that list? I'm OK with that, the application developer will be the person with the most knowledge of the app specifics.

Since you mention `setAdaptor` being called a second time I guess that means we will have the `ServiceAdaptor` code loaded by default for code bundles?

The proposal mentions we want to reduce service injection, it doesn't say why we would though, what are the reasons?

From what I remember this isn't quite what was proposed during our meeting, that proposal still relied on the module system and would have had the same caching issues the current approach has. From what I can see this approach uses the fact that BR matches on strings like `require(` to still bundle the correct service classes, is that correct? That's a clever way around the issue of not wanting to scan for magic strings but I wonder what would happen when we move to ES6 modules? ES6 modules won't require a server step that scans for `require(`.

Also non zero argument services don't seem to be addressed?

------------------------------------------------------------------------------------------

The approach that we discussed during the meeting on these issues is still worth considering, in essence it's pure code and no configuration, the solution we originally had years ago. Application developers could access the services either as they do now or we could recommend a type based approach instead if we prefer it, deprecating the old approach.

```javascript
var ServiceRegistry = require('br/ServiceRegistry');
var ConfigService = require('caplin/services/ConfigService');

var configService = ServiceRegistry.getService(ConfigService);
// OR
var configService = ServiceRegistry.getService('caplin.config-service');
```

Application developers would register services to the `ServiceRegistry` programatically. 

As many applications will use the same bundles of services we could have modules that they import and which could register bundles of services for them. Some obvious bundles are `BRServices`, `CaplinServices`, `WebcentricServices`, `SLJSServices`, we already have something similiar in the configuration approach with `useGroups="caplin.sljs-service-aliases"`.

Using code would require no special consideration for non zero args constructor services or async provided services.

Although I think taking a `Promise` based approach for resolving when async provided services are ready for use seems sensible and it might be good for us to standardize on that. We can have a simple interface that such services implement that returns a `Promise` and the ordering of services is dealt with in code.

```javascript
class Service<T> {
  static createInstance(): Promise<T> {
  }
}
```

I realize this might create more work for application developers but it's clear and requires no special knowledge to understand, it's all standard code. Require/import, instantiate, register. They have complete power and flexibility when it comes to all those steps.

**Comment by dchambers** [comment]
Hi @briandipalma, comments in-line...

> Why is 1 a problem? Thee interface doesn't have to reflect the fact the implementation is provided async, that seems to be an implementation detail. How does it cause problems?

It meant you couldn't change your mind later when you discovered an implementation of a service that needs to be asynchronous that didn't previously need to be. I don't think it was a question of the interface itself, since the asynchronicity was in the provisioning of the service, not in the interface itself.

@hobojed knows the most about this.

> 2 sounds like a bug, can we not just fix it? Neither issue is clear to me so I'm probably missing something.

We had a bootstrapping issue because there were two/three ways of creating service, none of which were guaranteed to dove-tail with the others when there were service inter-dependencies for services provisioned in the different ways:

  1. Injected service.
  2. Automatically created service.
  3. Automatically created service (asynchronous).

I created an issue specifically for these bootstrapping issues (#1335), though unfortunately there isn't really any more detailed information then what's here. I'd be happy to talk through exactly how this can cause a problem a call tomorrow if that's helpful?

> I'd recommend against exporting instances as the default export, in effect they become singletons and you probably want to test the service class, maybe a well known symbol on the default export i.e. INSTANCE might be more appropriate?

> I'd recommend against exporting instances as the default export, in effect they become singletons and you probably want to test the service class, maybe a well known symbol on the default export i.e. INSTANCE might be more appropriate?

> ```js
module.exports = ConsoleLogService;
module.exports.INSTANCE = Promise.resolve(new ConsoleLogService());
```

> A similar pattern could be applied to remove the need for the double modules for ServiceRegistry and ServiceRegistryClass and other comparable instances.

We avoided the singleton problem precisely by having two modules &mdash; `br/ServiceRegistry` (instance) and `br/ServiceRegistryClass` (class). This time around I was planning to make it feel more idiomatic by having `br/registry` (instance) and `br/Registry` (class) modules.

Perhaps there is a better way, but I don't really like the `INSTANCE` accessor as it has a Java like feel to it, and doesn't feel particularly idiomatic.

> We could consider using a type base registry in preference to a string based one at this juncture. It would allow tooling support to provide typing support once we move to using something like Flow or TS.

> ```js
var ServiceRegistry = require('br/ServiceRegistry');
var ConfigService = require('caplin/services/ConfigService');

> var configService = ServiceRegistry.getService(ConfigService);
```

> This would allow us to exploit polymorphism in the typed JS dialects. Strict nominal typing might be problematic in popouts (cross realm checks failing) so we might want to have a duck typing fallback if a service type doesn't match. Although I realize the benefit of this might be quite minor and not worth the hassle so we can just continue with the string based approach.

I'm not sure I understand your point, but the current string based approach doesn't prevent you having strongly typed services, or the use of Flow or TS. Whereas, not having a string identifier that we can identify on the server using a regular expression would prevent us from recognizing that an alias is being used, and knowing which concrete class to send to the client.

> I guess that detecting when and where it is safe to convert libraries to use async registry access will be very tricky and would counsel against such changes, especially as I'm doubtful of the value of such conversions. Also debugging Promises when they fail can be annoying, although the dev tools are improving. Given that we might want to question supporting that extension.

I'd be interested to know more about this. It's not currently clear to me why it's tricky to know if you can safely convert a library. Specifically, if you can do it without changing the API and breaking backwards compatibility then the asynchronicity has been contained, and will not spread any further.

> I like the prefetch approach, it's unclear how the list of async provided services is kept in sync, I guess application developers will have to create and maintain that list? I'm OK with that, the application developer will be the person with the most knowledge of the app specifics.

Yep, the app developer has the responsibility here. The important thing here is that when a synchronous require can't be provided, we need to give a clear error message to the developer like:

```js
The 'log-service' service could not be provided synchronously; you will need to either prefetch it before your app starts, or change code to require it asynchronously.
```

> Since you mention setAdaptor being called a second time I guess that means we will have the ServiceAdaptor code loaded by default for code bundles?

Here I was thinking about how CaplinTrader will provide a different implemenation of the `service` provider to the one BRJS provides &mdash; the CaplinTrader implementation will need to handle `DelayedReadinessService` for example. My thinking here is that we could just use a setter that can be called multiple times so that it can be overridden.

> The proposal mentions we want to reduce service injection, it doesn't say why we would though, what are the reasons?

Because it creates bootstrapping issues. If I inject a service (service `A`) that depends on service `B`, then we should perform service injection before we initialize (e.g. `prefetch()`) the other services. On the other hand, if service `B` depends on a service `C` that is also injected, then now it becomes harder to satisfy the dependencies in a good order. The more inter-dependent services you have, the worse this problem becomes.

> From what I remember this isn't quite what was proposed during our meeting, that proposal still relied on the module system and would have had the same caching issues the current approach has. From what I can see this approach uses the fact that BR matches on strings like require( to still bundle the correct service classes, is that correct? That's a clever way around the issue of not wanting to scan for magic strings but I wonder what would happen when we move to ES6 modules? ES6 modules won't require a server step that scans for require(.

Yes, I think you're quite possibly right that things have may have changed since we had the last conference call. The three intervening weeks and the holiday in between may have meant that things came out differently when I finally got the chance to put things down on paper. My poor memory seems to have helped for once!

> Also non zero argument services don't seem to be addressed?

They are addressed in that all services are now provided by modules that can either return a promisified service instance, or a non-promisified service constructor &mdash; we could support non-promisified service instances too, but `Promise.resolve(...)` is so easy to type anyway. So, if you have a service constructor that requires arguments than it's your responsibility to create a module that constructs it with the correct arguments, and then point your alias to that module.

> The approach that we discussed during the meeting...

I'm going to responding at this point since I think some of my earlier responses affect the arguments you're putting forward. Let's circle back if you still think this is relevant after you've digested this response.

**Comment by briandipalma** [comment]
Will skip discussing issue 1 as @hobojed has talked to both @dchambers and I and explained that he fixed the bug causing it.

>Services that required non zero-arg constructors had to be injected using ServiceRegistry.registerService(), yet doing so could prevent the service registry from correctly resolving delayed readiness services.

From the description of #1335 it seems that the issue is due to the fact that `var srv = require('service!...')` would automatically create a service on loading the code bundle but the service might not be one that can be created automatically as it needs constructor arguments or is provided async. This leaves the modules which `require` that service with a reference to an uninitialized service or a load time error, is that correct?

The identifier doesn't have to be `INSTANCE`. It's more about discussing a pattern that avoids the two module side-effect. Although this is an implementation detail, of low importance.

The type based registry was a spur of the moment suggestion and I didn't explain it well. In essence we could use polymorphism to flow the type from the ServiceRegistry to the service instance.

```javascript
class ServiceRegistry {
	services: Map<any, any>;
 
	getService<T>(requestedService: { new(): T; }): T {
		return this.services.get(requestedService);
	}
}
 
class MyService {
	myMethod() {}
}
 
var sr = new ServiceRegistry();
var myService = sr.getService(MyService);
 
myService.myMethod();
myService.incorrect(); // Property 'incorrect' does not exist on type 'MyService'.
```

Note this only catches the error in TS, Flow has a few bugs around polymorphism. If you use `string` then you won't be able to flow the type information like this. Mind you it's a minor issue, easily fixed with typecasting. I just thought if we could prevent some future typecasting it would be nice.

>Whereas, not having a string identifier that we can identify on the server using a regular expression would prevent us from recognizing that an alias is being used, and knowing which concrete class to send to the client.

In the simple proposal you leave the application developer to register which service implementations they want to use thus removing the need to perform string searches.

>I'd be interested to know more about this. It's not currently clear to me why it's tricky to know if you can safely convert a library. Specifically, if you can do it without changing the API and breaking backwards compatibility then the asynchronicity has been contained, and will not spread any further.

You can prevent bubbling up the fact that an operation is triggering async code but you can't prevent interleaving of code that previously didn't. Granted this would not be something I'd expect to cause many issues, but if it does cause any they won't be deterministic, so we could struggle to recreate issues clients can. I think problems would be rare though and this is distracting from the issue so I'll widthdraw my concerns.

Can we clarify what

>when a synchronous require can't be provided

means? Does it mean `var service = require('service!...');` How is the error raised? When is it raised?

>Here I was thinking about how CaplinTrader will provide a different implemenation of the service provider to the one BRJS provides ... My thinking here is that we could just use a setter that can be called multiple times so that it can be overridden.

So the default service adaptor will always be loaded and an application will have to load & register the CaplinTrader SDK one? That's fine.

>The more inter-dependent services you have, the worse this problem becomes.

Would we not recommend injecting services only for non zero arg constructor services? Chains of those should be rare and unless they are async provided should be simple to deal with. It sounds like we need to document where it makes sense to inject service instances.

We should clarify if this approach is working due to the current bundling of code scanning for strings matching `require('service!` and if that's so how it would work with ES6 modules as they won't `require` scan.

>They are addressed in that all services are now provided by modules that can either return a promisified service instance, or a non-promisified service constructor

I think I understand, for services that require constructor arguments the application developer should create and inject them into the ServiceRegistry?

**Comment by dchambers** [comment]
HI @briandipalma,

I'm going to focus on a single point you are making, since this seems to be the most defining difference between what I proposed and what you are proposing:

> > Whereas, not having a string identifier that we can identify on the server using a regular expression would prevent us from recognizing that an alias is being used, and knowing which concrete class to send to the client.

> In the simple proposal you leave the application developer to register which service implementations they want to use thus removing the need to perform string searches.

I'd forgotten the essential differences that made me think at the time that you should create a counter proposal, and this is it -- performing IoC using code only. The proposal you make about automatically flowing type information when `getService()` is invoked is possible as a consequence of that design choice, and wouldn't be possible with the approach I'm proposing.

I think this is an interesting idea, but I think it needs to be further developed in a fully worked through counter proposal. In particular, I'd like to see lots of attention paid to the backwards compatibility issues with existing code, and how we would seek to keep boiler plate to a minimum.

@adamshone / @bit-shifter / @hobojed / @janhancic / @james-shaw-turner: what do you guys think?

**Comment by briandipalma** [comment]
I agree on the similarties point. I'm currently fleshing out some ideas around the non string based approach. It's mostly synthesizing the ideas in the issue so far and ideas from our current approach into a summary as I start to lose track of how everything works when it's spread out over so many comments.

**Comment by james-s-turner** [comment]
Going back to issue 1
> synchronously provided services had to implement DelayedReadinessService, yet it was not possible for the original service interface author to know whether any future implementations might end up needing to be provided asynchronously."

Any service must know whether services it depends upon are synch or asynch so it knows when it can continue. Indeed the `DelayedReadiness` recognizes this by requiring the services to implement the method `getServiceDependencies()` that returns an array of other  `DelayedReadiness` services it depends upon.  Therefore if i'm developing `serviceA` that depends on `serviceB` and someone changes `serviceB` to become a delayed readiness service then `serviceA` will break.

**Comment by dchambers** [comment]
Hi @briandipalma,

I was talking to @james-shaw-turner and was reminded about some of the details of the potential problems I was alluding to with programmatic IoC:

**Boilerplate**:
You have to register every service that your code depends on, including any transitive dependencies. Worse still, you have to understand the full transitive dependency graph so that you can inject everything in the correct order.


**Backwards Compatibility Issues**:
At present it's possible for our library code to depend on a new service we introduce without breaking every app, workbench & test that our clients have written, but with programmatic IoC it's not.

**A Possible Solution?**:
You might suspect that you can solve both problems (_Boilerplate_ & _Backwards Compatibility Issues_) by having a class that helps to register all of the defaults you might want. Unfortunately this doesn't work very well since if you depend on such a helper class, than you will effectively depend on all of the default classes it refers to, even though you might only use a subset of these dependencies.

These were issues that led us to the solution we ended up with in the first place, and which continue to be reasons why our non programmatic approach is still the correct one.

**Comment by james-s-turner** [comment]
If I understand correctly  we boil down to: 
1. In Brian's approach the developer is responsible for populating the service registry programmatically at start up. (and as Dom states therefore has to understand the service dependency graph)
2. In Doms approach: Lazily load the registry by alias name, returning a Promise to allow for asynchronous retrieval from the registry. Additionally allow pre-loading of services so they can be retrieved synchronously (to aid backward compatibility, and possibly to reduce start time of the app)

**Comment by briandipalma** [comment]
Hi @james-shaw-turner 

Regarding issue 1. Which was

>Asynchronously provided services had to implement DelayedReadinessService, yet it was not possible for the original service interface author to know whether any future implementations might end up needing to be provided asynchronously.

I'm sorry about not providing details on the conversation that @hobojed had with @dchambers and me, it keeps everyone else following the issue in the dark. It appears that the bug that was fixed caused side effects such as the ones you describe

>Any service must know whether services it depends upon are synch or asynch so it knows when it can continue. Indeed the DelayedReadiness recognizes this by requiring the services to implement the method getServiceDependencies() that returns an array of other DelayedReadiness services it depends upon. Therefore if i'm developing serviceA that depends on serviceB and someone changes serviceB to become a delayed readiness service then serviceA will break.

Reading that description is seems that this is an issue introduced by working around the bug.

The bug was that a `DelayedService` provided a list of service dependencies and the `DelayedService` resolving logic would wait for all of those services to be constructed before proceding to construct the dependent service. The problem was that the service dependencies might not all be `DelayedService` ones and so might never resolve! The fix was to filter out sync services from the service dependencies list.

It looks like someone worked around this bug by altering the list of service dependencies for `serviceA` to not include `serviceB` but now the bug is fixed that work around can be removed.

**Comment by james-s-turner** [comment]
OK some light (i hope)!

That "fix" copes with the situation where you return a list of supposed async dependent services  but a dependency is actually synchronous. However there will be a failure if a  dependent async service is omitted from the list of dependent delayed readiness services. The bottom line is that a service needs to understand the async nature of its' dependencies.

**Comment by dchambers** [comment]
> The bottom line is that a service needs to understand the async nature of its' dependencies.

And they can't be changed without breaking backwards compatibility.

**Comment by briandipalma** [comment]
>However there will be a failure if a dependent async service is omitted from the list of dependent delayed readiness services.

That sounds like a mistake in the documentation of the `DelayedReadiness` `getServiceDependencies` method, how can we expect the `DelayedReadiness` logic to work unless you provide a complete list of service dependencies? The implementation already expects a mixed list of `DelayedReadiness` services and sync services and handles them just fine.

**Comment by james-s-turner** [comment]
Thanks @briandipalma, the explicitly (wrong) documentation caught me out !

> Only other <code>caplin.core.DelayedReadinessService</code>s should be specified.

So if a service returns a list of ALL its dependent services, then the registry can work out which are async services then it can "do the right thing". As long as there is a detection mechanism. Which we have in CT because of the DelayedReadinessService interface.

**Comment by briandipalma** [comment]
Nope, @james-shaw-turner, thank you. I was totally wrong about what I thought the documentation and code (that I probably wrote so this whole issue 1 is my fault) was saying and doing. It's only because you brought it up that I went and checked and saw that it was wrong.

**Comment by dchambers** [comment]
Having thought about this some more, I've come to the conclusion that:

  1. We should drop the idea of an asynchronous `require()`, and keep `require()` synchronous as it is now.
  2. Instead of having apps that have fine grained async service dependencies, that we should instead allow panels (implementing the `Component` interface) to start as soon as they can ensure their dependencies have been met, where some panels will start earlier than others.

These changes should vastly simplify the code people end up writing. As a consequence of this we would need to:

1. Make it the responsibility of the app developers to ensure dependencies are prefetched, for example:
  1. by using `registry.fetchAll()` to pre-fetch all service dependencies.
  2. by using `registry.fetch()` to fetch the minimal set of dependencies the app container needs, and then using further invocations of `registry.fetch()` for each panel within the app, so that panels can start to render as soon as they have the services they depend on.
2. Require that services that depend on other services make those dependencies known using a `dependentServices` property, and don't start requiring those services until the service constructor is invoked.
3. Encourage developers to associate their service aliases with a new `Service` interface so that a `registry.fetchAll()` method can be offered for apps that aren't interested in letting some components start before all of the services are ready.

For example, here is an example of what a service might look like:

```js
var registry = require('br/registry');
var service1, service2;

function MyService() {
	service1 = registry.require('service!service1');
	service2 = registry.require('service!service2');
}

MyService.dependentServices = ['service!service1', 'service!service2'];

module.exports = MyService;
```

where the `service` adaptor would automatically promisify `MyService`, using `fetch()` to ensure that all of the dependent services (specified in `dependentServices`) had first been resolved.

People needing a service module so they can define the constructor arguments would additionally need to write some code like this:

```js
var registry = require('br/registry');
var MyService = require('./MyService');

module.exports = registry.fetch(MyService.dependentServices).then(function() {
	return Promise.resolve(new MyService('some-arg'));
});
```

@james-shaw-turner, are you happy with this?

**Comment by briandipalma** [comment]
It might be helpful to specify the classes of services we have. This should make it easier for us to talk and reason about services. It should also help when developing utilities that can create, in the correct order, services.

It looks like we have several different classes of services. We have

`SimpleService`

* Depends on 0 services for correct execution.
* Zero arg constructor.
* Can be initialized in a sync manner.

`ParametricInitializationService`

* Depends on 0 services for correct execution.
* Non zero arg constructor.
* Can be initialized in a sync manner.

`DependentService`

* Depends on 1 or more services for correct execution.
* Zero arg constructor.
* Can be initialized in a sync manner.

`DependentParametricInitializationService`

* Depends on 1 or more services for correct execution.
* Non zero arg constructor.
* Can be initialized in a sync manner.

`AsyncInitializationService`

* Depends on 0 services for correct execution.
* Zero arg constructor.
* Require async operation to complete initialization.

`AsyncParametricInitializationService`

* Depends on 0 services for correct execution.
* Non zero arg constructor.
* Require async operation to complete initialization.

`DependentAsyncInitializationService`

* Depends on 1 or more services for correct execution.
* Zero arg constructor.
* Require async operation to complete initialization.

`DependentAsyncParametricInitializationService`

* Depends on 1 or more services for correct execution.
* Zero arg constructor.
* Require async operation to complete initialization.

Some of these classes of services we can create for the `ServiceRegistry` user but some we can't. Depending on the class of service the implementer will have to provide certain properties or methods on the implementation i.e. the `dependentServices` array @dchambers mentioned for dependent services or a static `createService(): Promise<T>` method on async initialization ones.

We still haven't address the question of how the string based approach works, I'm going to presume that it works due to the string `require('service!...'` still being used as the method signature. This leaves the question of how will this work when using ES6 modules that won't have any string searches carried out on them?

**Comment by dchambers** [comment]
Hi @briandipalma,

Regarding ES6 support, I'm imagining that will work by having BRJS spit out ES6 modules rather than CommonJS modules. This will be a trivial change for us to implement, that will more or less look like this:

  1. Change relevant `Asset` plug-ins to use ES6 module notation instead of invoking the `define()` method of [browser-modules](https://github.com/BladeRunnerJS/browser-modules).
  2. Delete [browser-modules](https://github.com/BladeRunnerJS/browser-modules).

Of course, we won't be able to do any of this stuff until we no longer have to support non evergreen versions of IE, which may never happen since I still don't think it's possible to prevent administrators from preventing IE from upgrading.

**Comment by briandipalma** [comment]
>Regarding ES6 support, I'm imagining that will work by having BRJS spit out ES6 modules rather than CommonJS modules.

I don't understand this...if I'm using ES6 modules I don't see what CJS has to do with it. What I'm saying is that ES6 modules will not have any `require` string searching operating on them which means we won't have a way of identifying service usage.

**Comment by dchambers** [comment]
> I don't understand this...if I'm using ES6 modules I don't see what CJS has to do with it. What I'm saying is that ES6 modules will not have any require string searching operating on them which means we won't have a way of identifying service usage.

Right, they have `import` statements that we'll have to start scanning for, and to remain backwardly compatible we'll need to scan for both `require()` and `import`.

**Comment by briandipalma** [comment]
>Right, they have import statements that we'll have to start scanning for, and to remain backwardly compatible we'll need to scan for both require() and import.

But we seperate code by different types, so namespace code only has namespace trie based text searches, CJS uses `require` and with ES6 it would have `import`. The reason for using `require` in CJS code was to keep it pure and not add magic string searches to CJS code. If we are happy to add a string search type that isn't pure ES6 to ES6 then why not just use the trie based search on CJS code and save us a lot of hassle? We can continue using `ServiceRegistry` as we've been doing for years now.

**Comment by briandipalma** [comment]
Also ES6 modules don't need to be loaded by BR, they can load themselves. There is also the fact that ES6 modules aren't lexically scoped so you couldn't just bundle up some extra modules like we currently do. Concating them will break them.

**Comment by dchambers** [comment]
Here are some of the negative things about namespaced style code that we were able to address when we added support for CommonJs style code:

  1. We were writing code in a way that was no longer idiomatic.
  2. We couldn't fail fast when somebody incorrectly referred to a module that didn't exist.
  3. We had to process each file with a trie which is marginally slower.
  4. We had to re-scan *all* namespaced files if any files were added, removed or renamed (effectively, when any change is detected in a `src` directory) since the trie may have changed.

I'm not sure which part of what I'm suggesting would not be _pure_ ES6. I'm merely suggesting that the server (BRJS) would also be able to determine what was going to be loaded on the client, since it will need to know this so it can ensure the right things are bundled or served.

**Comment by briandipalma** [comment]
Pure ES6 is the modules loading themselves. As I've already mentioned you can't bundle ES6 modules, it breaks them. They must request what they use you can't just concat extra code to the end of some modules.

**Comment by dchambers** [comment]
Yes, we can do that, but we'll still need to know:

  1. Which aliases map to which source modules.
  2. Which non JavaScript assets will need to be loaded to support the JavaScript modules that will be loaded.

To do this, we will still need to scan the code (be it ES6, CommonJS or Namespaced) to determine the dependencies. I know things like [webpack](http://webpack.github.io/) choose to convert all assets to JavaScript so they can be explicitly required, but we'll need to continue scanning the code on the server so we can support the apps and libraries people have already written.

**Comment by james-s-turner** [comment]
@briandipalma  I'm confused by your statement "you can't bundle ES6 modules". Isn't this exactly what happens now if you use Browserify and  Babel.  Also I found the following quote from:
http://guybedford.com/practical-workflows-for-es6-modules

> There has been some complaint about the fact that the ES6 Module Loader provides no native bundling format. But this becomes clear when understood in terms of the time scales of adoption for the spec.
> The way bundling will be enabled is through improvements at the protocol level such as SPDY and HTTP/2 which allow lots of small modules to be sent with similar performance to sending an entire file bundle.

Just because  ES6 provides no "native bundling format", I don't see how that means you cant bundle. As I read that, we can quite happily bundle ES6 modules until HTTP2 makes it irrelevant. Am I missing something?

**Comment by briandipalma** [comment]
>Which aliases map to which source modules.

As previously mentioned we can do this in code instead, it's straightforward.

>Which non JavaScript assets will need to be loaded to support the JavaScript modules that will be loaded.

The ES6 module loader pipeline allows you to translate non JS assets so you can still use ES6 modules to load these assets. This means there won't be a need to scan code to load it and its resources.

>we'll need to continue scanning the code on the server so we can support the apps and libraries people have already written.

I agree that for older codebases we would need to do that but we are talking about a new API and workflow for services. We should be building it to be future compatible.

**Comment by briandipalma** [comment]
> I'm confused by your statement "you can't bundle ES6 modules". Isn't this exactly what happens now if you use Browserify and Babel.

To clarify; what I mean is that you can't securely take X random ES6 modules and concatenate them. By bundling I mean concatenation.

It's important to not what @guybedford says in that blog post

>The way bundling will be enabled is through improvements at the protocol level

The important word being `protocol` - bundling will not happen at the source code level. You can have your server push ES6 modules to the browser but just because they are pushed to the client doesn't mean they will be executed or used. The ES6 source code has to import and make use of them!

**Comment by dchambers** [comment]
> As previously mentioned we can do this in code instead, it's straightforward.

Straightforward, but with massive downsides.

> The ES6 module loader pipeline allows you to translate non JS assets so you can still use ES6 modules to load these assets. This means there won't be a need to scan code to load it and its resources.

I'd heard about this, but I thought it was more of a hack like what webpack does. Surely this can only work if, like webpack, you require that all assets are explicitly required? Given that's not true for us, doesn't that mean we'll always need to scan on the server while we want to maintain backwards compatibility?

> I agree that for older codebases we would need to do that but we are talking about a new API and workflow for services. We should be building it to be future compatible.

I'm confused, are you suggesting that we have two modes of working depending on whether any of the libraries or code within the app doesn't explicitly require it's dependencies?

**Comment by dchambers** [comment]
@briandipalma, BTW, are you aware that we're doing an NPM spike soon so that people can start to consume NPM libraries more easily, but also so that NPM can be used for pure library development? I'm imagining that this is how people wanting to write completely new code bases will begin to move.

**Comment by dchambers** [comment]
@james-s-turner, yes @briandipalma is mostly right in that ES6 modules must be 'loaded' by the browser. There are hooks that provide flexibility in determining which URL to get them from, and my understanding is that it's even possible to intervene and not get them from the provided URL at all, and take them from a local bundle, but that's not how ES6 would normally be used.

> The way bundling will be enabled is through improvements at the protocol level such as SPDY and HTTP/2 which allow lots of small modules to be sent with similar performance to sending an entire file bundle.

The point here is that having the browser make hundreds of little requests won't be a performance problem by the time ES6 is widely deployed since HTTP/2 will be widely deployed by then anyway, and so the 'bundling' will effectively occur at the protocol level.

**Comment by briandipalma** [comment]
>Straightforward, but with massive downsides.

We might want to state these downsides so that we can analyze their size and if we can fix or reduce them. One of them was mentioned by @james-s-turner already.

>the developer is responsible for populating the service registry programmatically at start up. (and as Dom states therefore has to understand the service dependency graph)

There are ways to fix this issue. If we use the typed approach instead of the string approach. When you use the typed approach you need to import the service interface. That module can register itself with the registry as expecting an implementation to be provided for it. So on app load the service registry would log an error if the developer hasn't registered an implementation.

>I'd heard about this, but I thought it was more of a hack like what webpack does.

Nope it's the `translate` hook part of the `Loader` spec.

>Surely this can only work if, like webpack, you require that all assets are explicitly required?

Right you need to explictly declare your dependencies. It makes it much easier for developers to understand a codebase, one of the big pluses of modules.

>Given that's not true for us, doesn't that mean we'll always need to scan on the server while we want to maintain backwards compatibility?

Yes, for older codebases you would still need to scan for dependencies. We can use the code approach to registering services even in a namespaced code base, it's not hostile to being upgraded to.

>I'm confused, are you suggesting that we have two modes of working depending on whether any of the libraries or code within the app doesn't explicitly require it's dependencies?

It's like what we have with the current approach where if you have namespaced code or if you only have CJS code. It's future facing. We support old codebases but want to allow new codebases to drop practices that were previously required but no longer are.

**Comment by dchambers** [comment]
> We might want to state these downsides so that we can analyze their size and if we can fix or reduce them. One of them was mentioned by @james-s-turner already.

Ah, I previously stated the downsides in [this comment](https://github.com/BladeRunnerJS/brjs/issues/1128#issuecomment-95168684).

> Right you need to explictly declare your dependencies. It makes it much easier for developers to understand a codebase, one of the big pluses of modules.

Yes, I can see the value of doing this, it's just that we have to support lots of code that doesn't do this.

> Yes, for older codebases you would still need to scan for dependencies. We can use the code approach to registering services even in a namespaced code base, it's not hostile to being upgraded to.

Yes, but the downsides make this a bad idea IMO.

 > It's like what we have with the current approach where if you have namespaced code or if you only have CJS code. It's future facing. We support old codebases but want to allow new codebases to drop practices that were previously required but no longer are.

I don't think it is like that. CommonJS and namespaced code can be mixed and matched, but if a non-scanning ES6 project ever depended on an existing library then the entire code base would need to be scanned. We would also have two very different code paths, one of which would almost never be used, and just adding support for non-scanning ES6 would put a huge strain on our code-base, since it so different to what we have now.

Given that we're planning to start steering people towards NPM for library creation in the future, and given that our ability to drop support for browsers that don't support ES6 natively is a long way off, I'd suggest that any aspirations for non-scanning ES6 support are targeted for future apps that are comprised of NPM micro-libraries, and where we re-appraise how this is done based on the tools that will undoubtedly be created for the NPM community closer to the time.

**Comment by briandipalma** [comment]
>Boilerplate:
You have to register every service that your code depends on, including any transitive dependencies. Worse still, you have to understand the full transitive dependency graph so that you can inject everything in the correct order.

As previously mentioned you can deal with the dependency graph by requiring service interfaces to register themselves. We can provide modules that register commonly used bundles of service implementations, clients can use them if they find them convenient.

>Backwards Compatibility Issues:
At present it's possible for our library code to depend on a new service we introduce without breaking every app, workbench & test that our clients have written, but with programmatic IoC it's not.

This is only true if there is a `defaultClass` that makes sense to provide, when there isn't it also breaks. This is a manageable issue, with either the programmatic or string search based approach. Update app, load it, see error that a service hasn't been registered, register it. Not an onerous workflow for an issue that happens rarely.

>Unfortunately this doesn't work very well since if you depend on such a helper class, than you will effectively depend on all of the default classes it refers to, even though you might only use a subset of these dependencies.

If the helper bundles aren't acceptable it's a simple matter to create your own module. Both proposals expect the developer to create modules for services when they have that need. The string based approach does not eliminate the requirement to carry out work by the app dev.

Given the fact that the string based approach is future hostile it's difficult to see how it can be reasonably chosen as the soundest long term approach.

**Comment by dchambers** [comment]
Hi @briandipalma, sorry for the delay, I was interviewing...

> As previously mentioned you can deal with the dependency graph by requiring service interfaces to register themselves.

Can you please provide some more information about how the self registering interfaces would work to help with any dependency graph problems?

**Comment by briandipalma** [comment]
>Can you please provide some more information about how the self registering interfaces would work to help with any dependency graph problems?

When using the typed service registry you would access a service via its interface so you'd write code like this.

```javascript
var IMyService = require('caplin/lib/IMyService');
var serviceRegistry = require('br/ServiceRegistry');

...
var myServiceImpl = serviceRegistry.getService(IMyService);
```

This results in the service interface `IMyService` being loaded, and allows the service interface author to add registration code for `IMyService` to the `ServiceRegistry`.

```javascript
var serviceRegistry = require('br/ServiceRegistry');

function IMyService() {}

serviceRegistry.registerServiceInterface(IMyService);
```

When the service registry starts it verifies that an implementation is registered for every service interface. Any that aren't raise errors.

------------------------------------------------------------------------------------

Aside: A typed approach will allow us to more easily explore using the new DI library that is being created for Angular 2. They have moved from their old string based one to a typed based one due to various issues they had with the string based one.

```javascript
import {Inject} from 'di';
import {CoffeeMaker} from './coffee_maker/coffee_maker';
import {Skillet} from './skillet';
import {Stove} from './stove';
import {Fridge} from './fridge';
import {Dishwasher} from './dishwasher';

@Inject(CoffeeMaker, Skillet, Stove, Fridge, Dishwasher)
export class Kitchen {
  constructor(coffeeMaker, skillet, stove, fridge, dishwasher) {
    ...
  }
}
```

There is no need to use Decorators to use the library. You can imagine using the framework to provide services and as a way for our clients to more cleanly customize our code. This is influenced by Ninject and Guice which both eschew the older string based approach for a type/code based approach.

**Comment by dchambers** [comment]
Hi @briandipalma,

This has been a really interesting conversation, and I think you've really shown us what the future will look like. Sorry if I get some of the details wrong, but I think I can roughly summarize it like this:

  * HTTP/2 is used instead of bundling.
  * ES6 modules are used instead of CommonJS.
  * All assets are explicitly required, where non JavaScript assets are transformed into JavaScript using the ES6 'transform' hook.
  * No dependency analysis occurs on the server since the client just requests modules as it executes the code.
  * We can use programmatic IoC instead of aliases because:
    1. It becomes okay to refer to modules that have defaults for classes that we may not use without accidentally causing those modules to be loaded.
    2. We can safely introduce new services without breaking existing apps once it becomes possible for apps to refer to helper classes without loading too much stuff.
  * The service registry could potentially be written in a way that makes it possible for TypeScript to automatically 'flow' the type back into the variable that will receive the service.

However, while this stuff is all great, it depends on some technologies that are at least 5 years out in terms of when we'd realistically be able to make use of it in all the browsers we'd need to support. And, although it would be possible to begin to steer this way now, it currently comes with large downsides because:

  1. We have to scan the code on the back-end, and therefore need strings we can recognize, so we can bundle the code so the app is performant on today's browsers.
  2. We can't use programmatic IoC while we scan the code rather than evaluate it, since it would cause too much code to be loaded, or would cause breakage as new services are introduced.
  3. We have to support code that has implicit dependencies on non-JavaScript assets.
  4. We have to support code that has been written to depend on our existing IoC mechanism.

Therefore, I think we need an approach that steers as close in the 'right' direction as possible without doing things that will actually harm the experience today, and for the foreseeable future. Since this a profound and fundamental change, I plan to create a wiki page that attempts to better explain the final solution, so it can be further socialized among the teams before we commit to it.

Anyway, thanks so much for all of the excellent feedback and re-education you've provided.

**Comment by briandipalma** [comment]
I'm talking about services, nothing else. We can keep aliases and scanning and bundling etc. You asked for my feedback due to the issues we've encountered with services being loaded through the module system. The simplest, most flexible and future proof approach is a programmatic approach as opposed to a magic string approach.

>We have to scan the code on the back-end, and therefore need strings we can recognize, so we can bundle the code so the app is performant on today's browsers.

I'm not saying we stop doing that, we just no longer expect that to be the way people get their service implementations loaded.

>We can't use programmatic IoC while we scan the code rather than evaluate it, since it would cause too much code to be loaded, or would cause breakage as new services are introduced.

Developers are free to not use the service bundle modules and registering service interfaces means we can error if an interface implementation is not provided.

>We have to support code that has implicit dependencies on non-JavaScript assets.

Not an issue for services.

>We have to support code that has been written to depend on our existing IoC mechanism.

We can continue to do so. There is nothing stopping people from updating to a programmatic approach from the currect require approach at their leisure.

**Comment by dchambers** [comment]
Unfortunately, allowing people to take a mix and match approach doesn't work. For example, the proposal in this issue is also trying to address the [service bootstrapping problem](https://github.com/BladeRunnerJS/brjs/issues/1335) caused by allowing services to both be injected and automatically created, yet allowing both styles of code means another five years of these types of problems, rather than fixing things now.

Additionally, programmatic IoC without native ES6 and HTTP/2 means regular backwardly incompatibility changes as services are introduced or changed, whereas at the moment we don't have that problem. Or, if we use helper classes to create sensible defaults for everything then all apps will end up loading all services. 

Finally, supporting both programmatic and declarative IoC mechanisms at the same time, means lots more bugs and testing for developers, since things that may work for apps that use one style, might not for apps that use the other.

And, for all of these downsides we won't get a single benefit until all of the browsers we have to support have native ES6 support **and** all of the servers we have to support support HTTP/2 **and** we have apps that make no use of any libraries that were created with pre ES6 code. At that point we have to wonder why such an app would use anything like BRJS or webpack, when these will have long become irrelevant, and given that the NPM community and code base will have migrated to some ES6 compatible alternative. Really, in terms of the changes that will have occurred by that point in time, we really can't assume anything about how people will be developing by then at all.

**Comment by janhancic** [comment]
This probably fits into this discussion.

Even though proper ES6 modules are years away. We will probably start writing ES6 code soon, with the help of a transpiler. We don't have to wait 5 years before we can start doing that. Wouldn't it then make sense to propose a solution that will be future proof?

**Comment by briandipalma** [comment]
>However, while this stuff is all great, it depends on some technologies that are at least 5 years out

So the plan is to start working on ES6 support for BR in the summer of 2020?

**Comment by dchambers** [comment]
> Even though proper ES6 modules are years away. We will probably start writing ES6 code soon, with the help of a transpiler. We don't have to wait 5 years before we can start doing that. Wouldn't it then make sense to propose a solution that will be future proof?

@janhancic, unfortunately, using a tranpiler like Babel doesn't make it possible to move to fine-grained client-driven module loading as it doesn't support `System.import()`; only native ES6 module and HTTP/2 support will do. Of course, we could use [SystemJS](https://github.com/systemjs/systemjs) to shim in the dynamic module loading support, but we'd still need all of the browsers we need to support to have HTTP/2 baked in, and that won't happen until the oldest browser we need to support is IE12.

**Comment by janhancic** [comment]
I was thinking more in the line of transpiling ES6 imports and stuff into CommonJS. So the browser would get the same things as it does now. The only difference would be that devs write in ES6.

**Comment by briandipalma** [comment]
>Of course, we could use SystemJS to shim in the dynamic module loading support, but we'd still need all of the browsers we need to support to have HTTP/2 baked in

The module loader polyfill doesn't require that your browser supports HTTP/2. It works just fine without it.

**Comment by dchambers** [comment]
> I was thinking more in the line of transpiling ES6 imports and stuff into CommonJS. So the browser would get the same things as it does now. The only difference would be that devs write in ES6.

Ah, okay, yes we'll definitely be able to do that, and the current design also allows the sending of untranspiled ES6 to browsers too. What we've been discussing here though is whether it's possible to have what @briandipalma termed as 'pure' ES6 modules, where there is no server component involved in dependency analysis, and everything is client driven.

**Comment by dchambers** [comment]
> The module loader polyfill doesn't require that your browser supports HTTP/2. It works just fine without it.

Functionally, yes, but with completely unacceptable performance over a latent connection.

**Comment by briandipalma** [comment]
>Functionally, yes, but with completely unacceptable performance over a latent connection.

Has support for creating bundles of your source code so that's not an issue either.

**Comment by janhancic** [comment]
I know. What I'm trying to say is, that if we are going to allow devs to write ES6 code that then we should come up with a solution that works with ES6.

It would be weird writing ES6 code (transpiled) but using a feature that would make it impossible to actually write valid ES6.

Hope that makes sense :)

**Comment by briandipalma** [comment]
>@briandipalma termed as 'pure' ES6 modules, where there is no server component involved in dependency analysis, and everything is client driven.

There's nothing pure about that it's just ES6 modules functioning normally.

**Comment by dchambers** [comment]
> I know. What I'm trying to say is, that if we are going to allow devs to write ES6 code that then we should come up with a solution that works with ES6.

> It would be weird writing ES6 code (transpiled) but using a feature that would make it impossible to actually write valid ES6.

@janhancic, I think you've got the wrong end of the stick here. Nobody is suggesting you won't be able to write ES6 modules. The conversation has been about allowing the use of programmatic IoC for registering services, but how that has serious downsides until browsers support ES6 modules and HTTP/2.

**Comment by james-s-turner** [comment]
Doing a bit more summarising ;-)
> So the plan is to start working on ES6 support for BR in the summer of 2020?

No I want us to start writing ES6 code as soon as practically possible. That will involve transpiling and bundling for the near to medium term. As Dom says once there is full web ecosystem support for ES6 and HTTP2 (and we have a strategy to cope with our legacy code) we won't need most of the features that BRJS provides and we can converge to what is "industry standard" at that time.

Doms' summary asserts that Brians approach isn't a practical solution, until we have HTTP2/ES6 native support nirvana. Brians' more recent comments assert there are solutions to the problems Dom forsees e.g.
> Functionally, yes, but with completely unacceptable performance over a latent connection.
>> Has support for creating bundles of your source code so that's not an issue either.

@briandipalma could you explain in more detail what all the moving parts are in your proposed solution. e.g. It may "have support" what is that support how do we use it?  We really need to understand the details and be confident  how everything will work before we can make a judgement on the best approach. What software exists off the shelf at the moment and what would we have to write? 

Please open a new issue with your proposal so we can discuss it in isolation (this thread is already very long !). Also give some thought to backwards compatibility and/or the scope of any changes we would need to make to existing code.

**Comment by dchambers** [comment]
From looking afresh I've just found that there is something called `System.register()` that allows ES6 modules to be bundled and pushed at the browser without needing to do anything quirky like overriding the 'load' hook on the module loader. More information here:

  * https://github.com/ModuleLoader/es6-module-loader/wiki/System.register-Explained

Not sure what the practical upshot of this is yet, but sharing early...

**Comment by dchambers** [comment]
It just occurred to me that the ES6 utility classes you'd need to write to support programmatic IoC would need to use `System.import()` rather than use static `import` statements, so they could prevent the classes being loaded until the registry knows that the concrete class for some type is actually needed. While fine for a pure ES6 + HTTP/2 browser set-up, this would mean that even transpilers like Babel wouldn't bundle these modules, and so a server would need to be made available to serve them dynamically if needed, causing massive performance problems as a cascade of transitive dependencies are then loaded as a series of requests over an HTTP/1.1 connection.

**Comment by briandipalma** [comment]
>No, I want us to start writing ES6 code as soon as practically possible. That will involve transpiling and bundling for the near to medium term.

Depends on the browsers that you have to support. It seems realistic to say that within a year we will have Fx/Chrome and Spartan all supporting the ES6 modules (they already support HTTP/2 and large parts of the ES6 spec). IE8 will no longer be supported by MS and IE9 and IE10 will only be supported on older Windows Server versions and Vista. Neither of which are popular desktop OSes. It's conceivable that new and current projects will be looking to support only the latest Chrome/Fx and IE11+/Spartan. We already have a team working on an app that only supports Chrome why force them and other teams to go through build steps and use special server side machinery when they might not need them?

>As Dom says once there is full web ecosystem support for ES6 and HTTP2 (and we have a strategy to cope with our legacy code) we won't need most of the features that BRJS provides and we can converge to what is "industry standard" at that time.

We have code that is still in use after 7 - 8 years which will probably continue to be used for years. We should be planning for a decade not for 5 years at best and depending on what a team has to support 2 or 3 years. We should be thinking about using standards right now not 5 years down the line when the cost to move to standards is much higher.

>@briandipalma could you explain in more detail what all the moving parts are in your proposed solution.

This issue is around services, we should focus on making services a future friendly feature. There is one major difference between what is currently available and used in CJS code and the programmatic approach and that is that service implementation classes are registered by the application creator and not by automatic means.

Every other detail can be exactly the same.

**Comment by dchambers** [comment]
This is now covered by #1591.
